[WIP] Translate _d_arraycatnTX to template

Vild opened the pull request:
druntime PR: https://github.com/dlang/druntime/pull/2648

These patches will lower `a ~ a` to `(S[][2] __cat2 = [arr1, arr1];) , _d_arraycatnTX(cast(S[][])__cat2)`.

JinShil commented:
> ```D
> // This line will assert the compile:
> //     dmd: dmd/backend/cod4.d:401: Assertion `cast(int)sz > 0' failed.
> S[] arr3 = arr1 ~ arr1 ~ arr1;
> ```

@WalterBright Could you please lend your expertise to help us solve the backend assertion failure?

JinShil commented:
> So I guess I need to mark my static-array as a scope type,, but I could not find how to do that.

I believe you should be able to do that with `whatever.storageClass |= STC.scope_`

jacob-carlborg commented on src/dmd/expressionsem.d:
```diff
@@ -9636,36 +9636,115 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
+                auto comma = exp.isCommaExp();
+                if (!comma)
+                    return false;
+                if (auto e2Call = comma.e2.isCallExp())
```
Not technically necessary but I suggest that you added curly braces to this `if`-statement.

jacob-carlborg commented on src/dmd/dinterpret.d:
```diff
@@ -4997,6 +4997,30 @@ public:
+
+                Expression ea = (*e.arguments)[0];
+
+                while(true) {
```
* Curly braces on their own lines
* Space after `while`

Vild committed:
Lower array concat to _d_arraycatnTX

Signed-off-by: Dan Printzell <xwildn00bx@gmail.com>

Vild committed:
Remove backend code for _d_arraycatnTX

Signed-off-by: Dan Printzell <xwildn00bx@gmail.com>

Vild committed:
Intercept and handle _d_arraycatnTX for CTFE

Signed-off-by: Dan Printzell <xwildn00bx@gmail.com>

Vild committed:
Implement check for _d_arraycatnTX in nogc.d

Signed-off-by: Dan Printzell <xwildn00bx@gmail.com>

Vild committed:
Added symbol to hello-profile.d.trace.def

Signed-off-by: Dan Printzell <xwildn00bx@gmail.com>

Vild committed:
dmd.mars.parseCommandLine: force error format to be a `const(char)*`

Without this the first argument turns into a `string`
src/dmd/mars.d(2174): Error: function
`dmd.mars.parseCommandLine.error(const(char)* format, const(char*) arg
= null)` is not callable using argument types `(string, const(char)*)`

Signed-off-by: Dan Printzell <xwildn00bx@gmail.com>

rainers commented on src/dmd/dinterpret.d:
```diff
@@ -5019,6 +5019,33 @@ public:
+
+                Expression ea = (*e.arguments)[0];
+
+                while (true)
```
You have used a simpler `while(ea.isCastExp()) ea = ea.isCastExp().e1;` elsewhere.

rainers commented on src/dmd/dinterpret.d:
```diff
@@ -5019,6 +5019,33 @@ public:
+                }
+
+                if (global.params.verbose)
+                    message("interpret  %s =>\n          %s", e.toChars(), e1.toChars());
```
Do not use message for your debug output, use versioned printf instead.

rainers commented on src/dmd/expressionsem.d:
```diff
@@ -9624,15 +9624,15 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 exp.type = tb1next.arrayOf();
             L2elem:
-                if (tb2.ty == Tarray || tb2.ty == Tsarray)
+                if (tb2.ty != Tarray && tb2.ty != Tsarray)
```
Why can you completely invert the condition here?

rainers commented on src/dmd/expressionsem.d:
```diff
@@ -9624,15 +9624,15 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
-                result = exp.optimize(WANTvalue);
-                return;
+                e = exp.optimize(WANTvalue);
+                goto Lrewrite;
```
Can you place the rewriting into a (local) function instead of having to use goto?

rainers commented on src/dmd/mars.d:
```diff
@@ -2171,7 +2171,7 @@ bool parseCommandLine(const ref Strings arguments, const size_t argc, ref Param
                     if (!flag)
                     {
-                        error("unknown JSON field `-Xi=%s`, expected one of " ~ jsonFieldNames, p + 4);
+                        error(cast(const(char)*)("unknown JSON field `-Xi=%s`, expected one of " ~ jsonFieldNames), p + 4);
```
I haven't looked closely at the actual rewrite, but if it needs this additional cast (probably concatenation moved from CT to RT) it will break quite some code,

RazvanN7 commented:
Closing as @teodutu is taking care of this.
